File: libshvcore/src/chainpack/rpcmessage.cpp
#include "rpcmessage.h"
#include "chainpackprotocol.h"
#include "metatypes.h"

#include <cassert>

namespace shv {
namespace core {
namespace chainpack {

namespace meta {
RpcMessage::RpcMessage()
	: Super("RpcMessage")
{
	m_keys = {
		{(int)Key::Method, {(int)Key::Method, "Method"}},
		{(int)Key::Params, {(int)Key::Params, "Params"}},
		{(int)Key::Result, {(int)Key::Result, "Result"}},
		{(int)Key::Error, {(int)Key::Error, "Error"}},
		{(int)Key::ErrorCode, {(int)Key::ErrorCode, "ErrorCode"}},
		{(int)Key::ErrorMessage, {(int)Key::ErrorMessage, "ErrorMessage"}},
	};
	m_tags = {
		{(int)Tag::RequestId, {(int)Tag::RequestId, "RequestId"}},
		{(int)Tag::RpcCallType, {(int)Tag::RpcCallType, "RpcCallType"}},
		{(int)Tag::ShvPath, {(int)Tag::ShvPath, "ShvPath"}},
	};
}

void RpcMessage::registerMetaType()
{
	static bool is_init = false;
	if(!is_init) {
		is_init = true;
		static RpcMessage s;
		meta::registerType(meta::ChainPackNS::ID, meta::RpcMessage::ID, &s);
	}
}

}

//==================================================================
// RpcMessage
//==================================================================
RpcMessage::RpcMessage()
{
	meta::RpcMessage::registerMetaType();
}

RpcMessage::RpcMessage(const RpcValue &val)
	: RpcMessage()
{
	assert(val.isIMap());
	m_value = val;
}

bool RpcMessage::hasKey(RpcValue::UInt key) const
{
	return m_value.toIMap().count(key);
}

RpcValue RpcMessage::value(RpcValue::UInt key) const
{
	return m_value.at(key);
}

void RpcMessage::setValue(RpcValue::UInt key, const RpcValue &val)
{
	assert(key >= meta::RpcMessage::Key::Method && key < meta::RpcMessage::Key::MAX);
	checkMetaValues();
	m_value.set(key, val);
}

RpcValue RpcMessage::metaValue(RpcValue::UInt key) const
{
	return m_value.metaValue(key);
}

void RpcMessage::setMetaValue(RpcValue::UInt key, const RpcValue &val)
{
	checkMetaValues();
	m_value.setMetaValue(key, val);
}

RpcValue::UInt RpcMessage::id() const
{
	if(isValid())
		return m_value.metaData().value(meta::RpcMessage::Tag::RequestId).toUInt();
	return 0;
}

void RpcMessage::setId(RpcValue::UInt id)
{
	checkMetaValues();
	checkRpcTypeMetaValue();
	setMetaValue(meta::RpcMessage::Tag::RequestId, id);
}

bool RpcMessage::isValid() const
{
	return m_value.isValid();
}

bool RpcMessage::isRequest() const
{
	return rpcType() == meta::RpcMessage::RpcCallType::Request;
}

bool RpcMessage::isNotify() const
{
	return rpcType() == meta::RpcMessage::RpcCallType::Notify;
}

RpcValue RpcMessage::shvPath() const
{
	return metaValue(meta::RpcMessage::Tag::ShvPath);
}

void RpcMessage::setShvPath(const RpcValue &path)
{
	setMetaValue(meta::RpcMessage::Tag::ShvPath, path);
}

bool RpcMessage::isResponse() const
{
	return rpcType() == meta::RpcMessage::RpcCallType::Response;
}

int RpcMessage::write(std::ostream &out) const
{
	assert(m_value.isValid());
	assert(rpcType() != meta::RpcMessage::RpcCallType::Undefined);
	return ChainPackProtocol::write(out, m_value);
}

meta::RpcMessage::RpcCallType::Enum RpcMessage::rpcType() const
{
	RpcValue::UInt rpc_id = id();
	bool has_method = hasKey(meta::RpcMessage::Key::Method);
	if(has_method)
		return (rpc_id > 0)? meta::RpcMessage::RpcCallType::Request: meta::RpcMessage::RpcCallType::Notify;
	if(hasKey(meta::RpcMessage::Key::Result) || hasKey(meta::RpcMessage::Key::Error))
		return meta::RpcMessage::RpcCallType::Response;
	return meta::RpcMessage::RpcCallType::Undefined;
}

void RpcMessage::checkMetaValues()
{
	if(!m_value.isValid()) {
		m_value = RpcValue::IMap();
		/// not needed, Global is default name space
		setMetaValue(meta::Tag::MetaTypeNameSpaceId, meta::ChainPackNS::ID);
		setMetaValue(meta::Tag::MetaTypeId, meta::RpcMessage::ID);
	}
}

void RpcMessage::checkRpcTypeMetaValue()
{
	meta::RpcMessage::RpcCallType::Enum rpc_type = isResponse()? meta::RpcMessage::RpcCallType::Response: isNotify()? meta::RpcMessage::RpcCallType::Notify: meta::RpcMessage::RpcCallType::Request;
	setMetaValue(meta::RpcMessage::Tag::RpcCallType, rpc_type);
}

std::string RpcMessage::toStdString() const
{
	return m_value.toStdString();
}

//==================================================================
// RpcRequest
//==================================================================
RpcValue::String RpcRequest::method() const
{
	return value(meta::RpcMessage::Key::Method).toString();
}

RpcRequest &RpcRequest::setMethod(RpcValue::String &&met)
{
	setValue(meta::RpcMessage::Key::Method, RpcValue{std::move(met)});
	checkRpcTypeMetaValue();
	return *this;
}

RpcValue RpcRequest::params() const
{
	return value(meta::RpcMessage::Key::Params);
}

RpcRequest& RpcRequest::setParams(const RpcValue& p)
{
	setValue(meta::RpcMessage::Key::Params, p);
	return *this;
}

//==================================================================
// RpcNotify
//==================================================================
void RpcNotify::write(std::ostream &out, const std::string &method, std::function<void (std::ostream &)> write_params_callback)
{
	RpcValue::MetaData md;
	md.setMetaTypeId(meta::RpcMessage::ID);
	md.setValue(meta::RpcMessage::Tag::RpcCallType, meta::RpcMessage::RpcCallType::Notify);
	ChainPackProtocol::writeMetaData(out, md);
	ChainPackProtocol::writeContainerBegin(out, RpcValue::Type::IMap);
	ChainPackProtocol::writeMapElement(out, meta::RpcMessage::Key::Method, method);
	ChainPackProtocol::writeUIntData(out, meta::RpcMessage::Key::Params);
	write_params_callback(out);
	ChainPackProtocol::writeContainerEnd(out);
}

//==================================================================
// RpcResponse
//==================================================================
RpcResponse::Error RpcResponse::error() const
{
	return Error{value(meta::RpcMessage::Key::Error).toIMap()};
}

RpcResponse &RpcResponse::setError(RpcResponse::Error err)
{
	setValue(meta::RpcMessage::Key::Error, std::move(err));
	checkRpcTypeMetaValue();
	return *this;
}

RpcValue RpcResponse::result() const
{
	return value(meta::RpcMessage::Key::Result);
}

RpcResponse& RpcResponse::setResult(const RpcValue& res)
{
	setValue(meta::RpcMessage::Key::Result, res);
	checkRpcTypeMetaValue();
	return *this;
}

RpcResponse::Error::ErrorType RpcResponse::Error::code() const
{
	auto iter = find(KeyCode);
	return (iter == end()) ? NoError : (ErrorType)(iter->second.toUInt());
}

RpcResponse::Error& RpcResponse::Error::setCode(ErrorType c)
{
	(*this)[KeyCode] = RpcValue{(RpcValue::UInt)c};
	return *this;
}

RpcResponse::Error& RpcResponse::Error::setMessage(RpcValue::String &&mess)
{
	(*this)[KeyMessage] = RpcValue{std::move(mess)};
	return *this;
}

RpcValue::String RpcResponse::Error::message() const
{
	auto iter = find(KeyMessage);
	return (iter == end()) ? RpcValue::String{} : iter->second.toString();
}

} // namespace chainpackrpc
}
} // namespace shv

File: python/chainpack/rpcmessage.py
import enum
import meta
from value import *


class RpcMessage():
	def __init__(s, val: RpcValue = None) -> None:
		if val == None:
			val = RpcValue(invalid_value, Type.INVALID)
		else:
			assert(val.isIMap());
		s._value: RpcValue = val;

	def hasKey(s, key: meta.RpcMessage.Key) -> bool:
		if s._value.type == Type.IMap:
			return key in s._value.toPython()
		else:
			return False

	def value(s, key: uint) -> RpcValue:
		return s._value._value[key]

	def setValue(s, key: uint, val: RpcValue):
		assert(key >= meta.RpcMessage.Key.Method and key < meta.RpcMessage.Key.MAX);
		s.checkMetaValues();
		s._value._value[key] = val

	def setMetaValue(s, key: int, val: RpcValue):
		s.checkMetaValues();
		s._value.setMetaValue(key, val);

	def id(s) -> uint:
		if s.isValid() and (meta.RpcMessage.Tag.RequestId in s._value._metaData):
			return s._value._metaData[meta.RpcMessage.Tag.RequestId].toUInt();
		return 0

	def setId(s, id: (uint, int)) -> None:
		s.checkMetaValues();
		s.checkRpcTypeMetaValue();
		s.setMetaValue(meta.RpcMessage.Tag.RequestId, RpcValue(uint(id)));

	def isValid(s) -> bool:
		return s._value.isValid()

	def isRequest(s) -> bool:
		return s.rpcType() == meta.RpcMessage.RpcCallType.Request

	def isNotify(s) -> bool:
		return s.rpcType() == meta.RpcMessage.RpcCallType.Notify

	def isResponse(s) -> bool:
		return s.rpcType() == meta.RpcMessage.RpcCallType.Response

	def write(s, out: ChainPackProtocol) -> int:
		assert(s._value.isValid());
		assert(s.rpcType() != meta.RpcMessage.RpcCallType.Undefined);
		return out.write(s._value);

	def rpcType(s) -> meta.RpcMessage.RpcCallType:
		rpc_id: int = s.id();
		has_method: bool = s.hasKey(meta.RpcMessage.Key.Method);
		if(has_method):
			if rpc_id > 0:
				return meta.RpcMessage.RpcCallType.Request
			else:
				return meta.RpcMessage.RpcCallType.Notify;
		if s.hasKey(meta.RpcMessage.Key.Result) or s.hasKey(meta.RpcMessage.Key.Error):
			return meta.RpcMessage.RpcCallType.Response;
		return meta.RpcMessage.RpcCallType.Undefined;

	def checkMetaValues(s) -> None:
		if not s._value.isValid():
			s._value = RpcValue({}, Type.IMap)
			s._value.setMetaValue(meta.Tag.MetaTypeId, meta.RpcMessage.ID);

	def checkRpcTypeMetaValue(s):
		if s.isResponse():
			rpc_type = meta.RpcMessage.RpcCallType.Response
		elif s.isNotify():
			rpc_type = meta.RpcMessage.RpcCallType.Notify
		else:
			rpc_type = meta.RpcMessage.RpcCallType.Request
		s.setMetaValue(meta.RpcMessage.Tag.RpcCallType, rpc_type);

	def method(s) -> str:
		return s.value(meta.RpcMessage.Key.Method).toString();

	def setMethod(s, met: str) -> None:
		s.setValue(meta.RpcMessage.Key.Method, RpcValue(met));

	def params(s) -> RpcValue:
		return value(meta.RpcMessage.Key.Params);

	def setParams(s, p: RpcValue) -> None:
		s.setRpcValue(meta.RpcMessage.Key.Params, p);


class RpcRequest(RpcMessage):
	def params(s) -> RpcValue:
		return s.value(meta.RpcMessage.Key.Params);

	def setParams(s, p: RpcValue):
		s.setValue(meta.RpcMessage.Key.Params, RpcValue(p));


class RpcResponse:
	pass
class RpcResponse(RpcMessage):

	class Key(enum.IntFlag):
		Code = 1,
		KeyMessage = 2

	class ErrorType(enum.IntFlag):
		NoError = 0,
		InvalidRequest = 1,
		MethodNotFound = 2,
		InvalidParams = 3,
		InternalError = 4,
		ParseError = 5,
		SyncMethodCallTimeout = 6,
		SyncMethodCallCancelled = 7,
		MethodInvocationException = 8,
		Unknown = 9

	class Error(dict):

		class Key(enum.IntFlag):
			Code = 1,
			Message = 2

		class ErrorType(enum.IntFlag):
			NoError = 0,
			InvalidRequest = 1,
			MethodNotFound = 2,
			InvalidParams = 3,
			InternalError = 4,
			ParseError = 5,
			SyncMethodCallTimeout = 5,
			SyncMethodCallCancelled = 6,
			MethodInvocationException = 7,
			Unknown = 8

		@classmethod
		def createError(cls, code, message):
			r = cls()
			r.setCode(code)
			r.setMessage(message)
			return r

		def code(s) -> ErrorType:
			if Key.Code in s:
				return s[Key.Code].toUInt();
			return ErrorType.NoError

		def setCode(s, c: ErrorType):
			s[s.__class__.Key.Code] = RpcValue(uint(c));
			return s;

		def setMessage(s, msg: str):
			s[s.__class__.Key.Message] = RpcValue(msg);
			return s

		def message(s) -> str:
			return s.get([Key.Message.toString()], "")

	def error(s) -> Error:
		return RpcResponse.Error(s.value(meta.RpcMessage.Key.Error).toPython());

	def setError(s, err):
		s.setValue(meta.RpcMessage.Key.Error, RpcValue(err, Type.IMap));
		s.checkRpcTypeMetaValue();

	def result(s) -> RpcValue:
		return s.value(meta.RpcMessage.Key.Result);

	def setResult(s, res: RpcValue) -> RpcResponse:
		s.setValue(meta.RpcMessage.Key.Result, RpcValue(res));
		s.checkRpcTypeMetaValue();

